lib/sycamore/sycamore/materialize.py

```diff
@@ -91,17 +94,45 @@ def __init__(
 
         super().__init__(child, **kwargs)
 
+        self._maybe_anonymous()
+
+    def _maybe_anonymous(self):
+        if self._root is None:
+            return
+        from pyarrow.fs import S3FileSystem
+
+        if not isinstance(self._fs, S3FileSystem):
+            return
+
+        try:
+            self._fs.get_file_info(str(self._root))
+            return
+        except OSError as e:
+            logging.warning(f"Got error {e} trying to get file info on {self._root}, trying again in anonymous mode")
+
+        fs = S3FileSystem(anonymous=True)
+        try:
+            fs.get_file_info(str(self._root))
+            self._fs = fs
+            self._fshelper = _PyArrowFsHelper(self._fs)
+            return
+        except OSError as e:
+            logging.warning(
+                f"Got error {e} trying to anonymously get file info on {self._root}. Likely to fail shortly."
+            )
+            return
+
     def prepare(self):
         """
         Clean up the materialize location if necessary.
         Validate that cleaning worked, but only once all materializes have finished cleaning.
         This protects against multiple materializes pointing to the same location.
         """
 
-        if self._will_be_source():
+        if self._root is None:
```

Comment:
Is there any meaning to this ordering change, or is it just to alphabetize?

Reply:
Check for no root before checking if a source since that makes more sense.